From 52ad4ec6b2b8c0facf785561df14ed5ccc6f5fcd Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Mon, 21 Jun 2021 11:43:05 +0200
Subject: [PATCH] PONRTSYS-9036: Set VLAN protocol only when it changed

When the VLAN protocol should not be modified in a VLAN modify action
take the VLAN protocol from the original inner or outer tag.
This will handle the VLAN protocol similar to the VLAN priority and tag.
When the VLAN push action is done assume that the VLAN protocol was
configured.
---
 .../net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
index 92bc9bd81971..85dd01903e64 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
@@ -309,6 +309,24 @@ static void vlan_action_vid_parse(struct net_device *dev,
 		*vid = CP_FROM_OUTER;
 }
 
+static void vlan_action_modify_proto_parse(struct net_device *dev,
+					   const struct tc_action *a,
+					   enum tc_flower_vlan_tag tag,
+					   int *tpid)
+{
+	u8 flags = tcf_vlan_push_flags(a);
+
+	if (flags & ACTVLAN_PUSH_F_PROTO) {
+		*tpid = ntohs(tcf_vlan_push_proto(a));
+		return;
+	}
+
+	if (tag == TC_VLAN_SINGLE_TAGGED)
+		*tpid = CP_FROM_INNER;
+	else if (tag == TC_VLAN_DOUBLE_TAGGED)
+		*tpid = CP_FROM_OUTER;
+}
+
 static int vlan_action_parse(struct net_device *dev,
 			     struct tc_cls_flower_offload *f,
 			     struct dp_act_vlan *act,
@@ -358,6 +376,7 @@ static int vlan_action_parse(struct net_device *dev,
 			act->act |= DP_VLAN_ACT_PUSH;
 
 			act->prio[act->push_n] = tcf_vlan_push_prio(a);
+			act->tpid[act->push_n] = ntohs(tcf_vlan_push_proto(a));
 			netdev_dbg(dev, "PUSH action detected, push_n: %d, act: %x\n",
 				   act->push_n + 1, act->act);
 		} else if (tcf_vlan_action(a) == TCA_VLAN_ACT_MODIFY) {
@@ -365,12 +384,13 @@ static int vlan_action_parse(struct net_device *dev,
 			act->pop_n++;
 			vlan_action_modify_prio_parse(dev, a, tag,
 						      &act->prio[act->push_n]);
+			vlan_action_modify_proto_parse(dev, a, tag,
+						      &act->tpid[act->push_n]);
 			netdev_dbg(dev, "MODIFY action detected, push_n: %d, pop_n: %d, act: %x\n",
 				   act->push_n + 1, act->pop_n, act->act);
 		}
 
 		vlan_action_vid_parse(dev, a, tag, &act->vid[act->push_n]);
-		act->tpid[act->push_n] = ntohs(tcf_vlan_push_proto(a));
 		act->dei[act->push_n] = 0;
 
 		netdev_dbg(dev, "PUSH action parsed, vid: %d, tpid: %x, prio: %d, dei: %d\n",
